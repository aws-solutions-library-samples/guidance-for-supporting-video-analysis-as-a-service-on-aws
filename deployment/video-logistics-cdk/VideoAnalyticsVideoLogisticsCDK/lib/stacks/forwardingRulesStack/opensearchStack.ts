import type { App } from 'aws-cdk-lib';
import { Domain } from 'aws-cdk-lib/aws-opensearchservice';
import { Duration, Fn, RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib';
import {getVLSearchDomainProps} from './utils';
import {
  AccountRootPrincipal,
  AnyPrincipal,
  ArnPrincipal,
  Effect,
  PolicyDocument,
  PolicyStatement,
  ServicePrincipal
} from 'aws-cdk-lib/aws-iam';
import { Key } from 'aws-cdk-lib/aws-kms';
import * as opensearchserverless from 'aws-cdk-lib/aws-opensearchserverless';
import { AWSRegion } from 'video_analytics_common_construct';
import {OPEN_SEARCH_SERVICE_NAME} from '../const';


export interface OpenSearchStackProps extends StackProps {
    region: AWSRegion;
    account: string;
}

export class OpenSearchStack extends Stack {
  public readonly opensearchEndpoint: string;

  constructor(scope: App, id: string, readonly props: OpenSearchStackProps) {
    super(scope, id, props);

    const vlControlPlaneBulkLambdaRoleArn =`arn:aws:iam::${props.account}:role/BulkInferenceLambdaRole`;
    const vlControlPlanePitCreationLambdaRoleArn =`arn:aws:iam::${props.account}:role/PitCreationLambdaRole`;

    const openSearchDomain = this.createOpenSearchDomain(props);

    this.opensearchEndpoint = openSearchDomain.domainEndpoint;

    openSearchDomain.addAccessPolicies(
        new PolicyStatement({
          actions: ['es:*'],
          effect: Effect.ALLOW,
          principals: [
            new ServicePrincipal(OPEN_SEARCH_SERVICE_NAME),
            new ArnPrincipal(vlControlPlaneBulkLambdaRoleArn),
        ],
          resources: [openSearchDomain.domainArn, `${openSearchDomain.domainArn}/*`]
        })
      );
    }

  private createOpenSearchDomain(props: OpenSearchStackProps): Domain {
    const domainId = `VALOpenSearchDomain`;
    const keyPolicy = new PolicyDocument({
      statements: [
        new PolicyStatement({
          actions: [
            'kms:List*',
            'kms:Describe*'
          ],
          principals: [new AnyPrincipal()],
          resources: ['*'],
          conditions: {
            ArnLike: {
              'aws:PrincipalArn': `arn:aws:iam::${this.account}:role/*`
            }
          }
        }),
        new PolicyStatement({
          actions: [
            'kms:CreateGrant'
          ],
          principals: [new AnyPrincipal()],
          resources: ['*'],
          conditions: {
            ArnLike: {
              'aws:PrincipalArn': `arn:aws:iam::${this.account}:role/*`
            },
            'StringEquals': {
              'kms:ViaService': `es.${props.region}.amazonaws.com`
            },
            'Bool': {
              'kms:GrantIsForAWSResource': 'true'
            }
          }
        })
      ]
    });
    const openSearchClusterKey = new Key(this, `VALDomainKey`, {
      enabled: true,
      alias: `DomainKey`,
      enableKeyRotation: true,
      admins: [new AccountRootPrincipal()],
      policy: keyPolicy
    });

    return new Domain(
        this,
        domainId,
        getVLSearchDomainProps(domainId.toLowerCase(), openSearchClusterKey)
    );
  }
}