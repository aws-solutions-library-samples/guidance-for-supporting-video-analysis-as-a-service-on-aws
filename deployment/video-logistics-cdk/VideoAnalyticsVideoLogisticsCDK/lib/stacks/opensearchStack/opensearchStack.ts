import type { App } from 'aws-cdk-lib';
import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import {
  AccountRootPrincipal,
  AnyPrincipal,
  ArnPrincipal,
  Effect,
  ManagedPolicy,
  PolicyDocument,
  PolicyStatement,
  Role,
  ServicePrincipal
} from 'aws-cdk-lib/aws-iam';
import { Key } from 'aws-cdk-lib/aws-kms';
import { Domain } from 'aws-cdk-lib/aws-opensearchservice';
import { AWSRegion } from 'video_analytics_common_construct';
import { OPEN_SEARCH_SERVICE_NAME } from '../const';
import { getVLSearchDomainProps } from './utils';


export interface OpenSearchStackProps extends StackProps {
    region: AWSRegion;
    account: string;
}

export class OpenSearchStack extends Stack {
  public readonly opensearchEndpoint: string;
  public readonly bulkInferenceLambdaRoleArn: string;

  constructor(scope: App, id: string, readonly props: OpenSearchStackProps) {
    super(scope, id, props);

    const bulkInferenceLambdaRole = new Role(this, 'BulkInferenceLambdaRole', {
      roleName: 'BulkInferenceLambdaRole',
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      description: 'Allows lambda to make a bulk request to open search'
    });

    bulkInferenceLambdaRole.addManagedPolicy(
      ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole')
    );

    const openSearchPolicy = new PolicyStatement({
      effect: Effect.ALLOW,
      resources: [
        `arn:aws:es:${this.region}:${this.account}:domain/valopensearchdomain`,
        `arn:aws:es:${this.region}:${this.account}:domain/valopensearchdomain/*`
      ],
      actions: ['es:ESHttpPost', 'es:ESHttpPut', 'es:ESHttpGet', 'es:ESHttpHead']
    });

    const kmsPolicy = new PolicyStatement({
      effect: Effect.ALLOW,
      resources: [`arn:aws:kms:${this.region}:${this.account}:key/*`],
      actions: ['kms:Decrypt', 'kms:Encrypt', 'kms:ReEncrypt*', 'kms:GenerateDataKey*']
    });

    bulkInferenceLambdaRole.addToPolicy(openSearchPolicy);
    bulkInferenceLambdaRole.addToPolicy(kmsPolicy);

    this.bulkInferenceLambdaRoleArn = bulkInferenceLambdaRole.roleArn;
    new CfnOutput(this, 'BulkInferenceLambdaRoleArn', {
      value: this.bulkInferenceLambdaRoleArn,
      exportName: 'BulkInferenceLambdaRoleArn',
    });

    const openSearchDomain = this.createOpenSearchDomain(props);

    this.opensearchEndpoint = openSearchDomain.domainEndpoint;

    openSearchDomain.addAccessPolicies(
        new PolicyStatement({
          actions: ['es:*'],
          effect: Effect.ALLOW,
          principals: [
            new ServicePrincipal(OPEN_SEARCH_SERVICE_NAME),
            new ArnPrincipal(bulkInferenceLambdaRole.roleArn),
        ],
          resources: [openSearchDomain.domainArn, `${openSearchDomain.domainArn}/*`]
        })
      );
    }

  private createOpenSearchDomain(props: OpenSearchStackProps): Domain {
    const domainId = `VALOpenSearchDomain`;
    const keyPolicy = new PolicyDocument({
      statements: [
        new PolicyStatement({
          actions: [
            'kms:List*',
            'kms:Describe*'
          ],
          principals: [new AnyPrincipal()],
          resources: ['*'],
          conditions: {
            ArnLike: {
              'aws:PrincipalArn': `arn:aws:iam::${this.account}:role/*`
            }
          }
        }),
        new PolicyStatement({
          actions: [
            'kms:CreateGrant'
          ],
          principals: [new AnyPrincipal()],
          resources: ['*'],
          conditions: {
            ArnLike: {
              'aws:PrincipalArn': `arn:aws:iam::${this.account}:role/*`
            },
            'StringEquals': {
              'kms:ViaService': `es.${props.region}.amazonaws.com`
            },
            'Bool': {
              'kms:GrantIsForAWSResource': 'true'
            }
          }
        })
      ]
    });
    const openSearchClusterKey = new Key(this, `VALDomainKey`, {
      enabled: true,
      alias: `DomainKey`,
      enableKeyRotation: true,
      admins: [new AccountRootPrincipal()],
      policy: keyPolicy
    });

    return new Domain(
        this,
        domainId,
        getVLSearchDomainProps(domainId.toLowerCase(), openSearchClusterKey)
    );
  }
}
