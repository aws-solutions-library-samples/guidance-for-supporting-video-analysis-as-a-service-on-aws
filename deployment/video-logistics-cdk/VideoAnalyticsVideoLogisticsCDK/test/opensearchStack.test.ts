import { createStackSnapshotTest } from "video_analytics_common_construct";
import { OpenSearchStack } from "../lib/stacks/opensearchStack/opensearchStack";

createStackSnapshotTest("OpenSearchStack", OpenSearchStack);
