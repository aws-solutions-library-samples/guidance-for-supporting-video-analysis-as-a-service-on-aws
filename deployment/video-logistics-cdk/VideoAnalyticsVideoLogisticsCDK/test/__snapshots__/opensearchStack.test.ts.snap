// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenSearchStack Snapshot test 1`] = `
{
  "Outputs": {
    "BulkInferenceLambdaRoleArn": {
      "Export": {
        "Name": "BulkInferenceLambdaRoleArn",
      },
      "Value": {
        "Fn::GetAtt": [
          "BulkInferenceLambdaRole4C07E6B3",
          "Arn",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleDefaultPolicyD28E1A5E",
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": Anything,
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleDefaultPolicyD28E1A5E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "kms:List*",
                "kms:Describe*",
                "kms:CreateGrant",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "VALDomainKey37626C63",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleDefaultPolicyD28E1A5E",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BulkInferenceLambdaRole4C07E6B3": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Allows lambda to make a bulk request to open search",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "BulkInferenceLambdaRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "BulkInferenceLambdaRoleDefaultPolicy57C8A9D3": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "es:ESHttpPost",
                "es:ESHttpPut",
                "es:ESHttpGet",
                "es:ESHttpHead",
              ],
              "Effect": "Allow",
              "Resource": [
                "arn:aws:es:us-east-1:123456789012:domain/valopensearchdomain",
                "arn:aws:es:us-east-1:123456789012:domain/valopensearchdomain/*",
              ],
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:kms:us-east-1:123456789012:key/*",
            },
            {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::video-analytics-image-upload-bucket-123456789012-us-east-1/*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BulkInferenceLambdaRoleDefaultPolicy57C8A9D3",
        "Roles": [
          {
            "Ref": "BulkInferenceLambdaRole4C07E6B3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "VALDomainKey37626C63": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "EnableKeyRotation": true,
        "Enabled": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": [
                "kms:List*",
                "kms:Describe*",
              ],
              "Condition": {
                "ArnLike": {
                  "aws:PrincipalArn": "arn:aws:iam::123456789012:role/*",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": "*",
            },
            {
              "Action": "kms:CreateGrant",
              "Condition": {
                "ArnLike": {
                  "aws:PrincipalArn": "arn:aws:iam::123456789012:role/*",
                },
                "Bool": {
                  "kms:GrantIsForAWSResource": "true",
                },
                "StringEquals": {
                  "kms:ViaService": "es.us-east-1.amazonaws.com",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:TagResource",
                "kms:UntagResource",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "VALDomainKeyAlias6C44A7F2": {
      "Properties": {
        "AliasName": "alias/DomainKey",
        "TargetKeyId": {
          "Fn::GetAtt": [
            "VALDomainKey37626C63",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "VALOpenSearchDomain7C5F7020": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "ClusterConfig": {
          "DedicatedMasterCount": 3,
          "DedicatedMasterEnabled": true,
          "DedicatedMasterType": "m6g.large.search",
          "InstanceCount": 3,
          "InstanceType": "c6g.large.search",
          "WarmCount": 3,
          "WarmEnabled": true,
          "WarmType": "ultrawarm1.medium.search",
          "ZoneAwarenessConfig": {
            "AvailabilityZoneCount": 3,
          },
          "ZoneAwarenessEnabled": true,
        },
        "DomainEndpointOptions": {
          "EnforceHTTPS": true,
          "TLSSecurityPolicy": "Policy-Min-TLS-1-0-2019-07",
        },
        "DomainName": "valopensearchdomain",
        "EBSOptions": {
          "EBSEnabled": true,
          "VolumeSize": 100,
          "VolumeType": "gp3",
        },
        "EncryptionAtRestOptions": {
          "Enabled": true,
          "KmsKeyId": {
            "Ref": "VALDomainKey37626C63",
          },
        },
        "EngineVersion": "OpenSearch_2.9",
        "LogPublishingOptions": {},
        "NodeToNodeEncryptionOptions": {
          "Enabled": true,
        },
      },
      "Type": "AWS::OpenSearchService::Domain",
      "UpdatePolicy": {
        "EnableVersionUpgrade": true,
      },
      "UpdateReplacePolicy": "Retain",
    },
    "VALOpenSearchDomainAccessPolicyCustomResourcePolicy0B24C3AD": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "es:UpdateDomainConfig",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "VALOpenSearchDomain7C5F7020",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "VALOpenSearchDomainAccessPolicyCustomResourcePolicy0B24C3AD",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "VALOpenSearchDomainAccessPolicyE52C5A6D": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "VALOpenSearchDomainAccessPolicyCustomResourcePolicy0B24C3AD",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"action":"updateDomainConfig","service":"OpenSearch","parameters":{"DomainName":"",
              {
                "Ref": "VALOpenSearchDomain7C5F7020",
              },
              "","AccessPolicies":"{\\"Statement\\":[{\\"Action\\":\\"es:*\\",\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"Service\\":\\"es.amazonaws.com\\",\\"AWS\\":\\"",
              {
                "Fn::GetAtt": [
                  "BulkInferenceLambdaRole4C07E6B3",
                  "Arn",
                ],
              },
              "\\"},\\"Resource\\":[\\"",
              {
                "Fn::GetAtt": [
                  "VALOpenSearchDomain7C5F7020",
                  "Arn",
                ],
              },
              "\\",\\"",
              {
                "Fn::GetAtt": [
                  "VALOpenSearchDomain7C5F7020",
                  "Arn",
                ],
              },
              "/*\\"]}],\\"Version\\":\\"2012-10-17\\"}"},"outputPaths":["DomainConfig.AccessPolicies"],"physicalResourceId":{"id":"",
              {
                "Ref": "VALOpenSearchDomain7C5F7020",
              },
              "AccessPolicy"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": false,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": {
          "Fn::Join": [
            "",
            [
              "{"action":"updateDomainConfig","service":"OpenSearch","parameters":{"DomainName":"",
              {
                "Ref": "VALOpenSearchDomain7C5F7020",
              },
              "","AccessPolicies":"{\\"Statement\\":[{\\"Action\\":\\"es:*\\",\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"Service\\":\\"es.amazonaws.com\\",\\"AWS\\":\\"",
              {
                "Fn::GetAtt": [
                  "BulkInferenceLambdaRole4C07E6B3",
                  "Arn",
                ],
              },
              "\\"},\\"Resource\\":[\\"",
              {
                "Fn::GetAtt": [
                  "VALOpenSearchDomain7C5F7020",
                  "Arn",
                ],
              },
              "\\",\\"",
              {
                "Fn::GetAtt": [
                  "VALOpenSearchDomain7C5F7020",
                  "Arn",
                ],
              },
              "/*\\"]}],\\"Version\\":\\"2012-10-17\\"}"},"outputPaths":["DomainConfig.AccessPolicies"],"physicalResourceId":{"id":"",
              {
                "Ref": "VALOpenSearchDomain7C5F7020",
              },
              "AccessPolicy"}}",
            ],
          ],
        },
      },
      "Type": "Custom::OpenSearchAccessPolicy",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
